x/ccv/provider/keeper/migration.go

```diff
@@ -0,0 +1,138 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	providertypes "github.com/cosmos/interchain-security/v3/x/ccv/provider/types"
+	ccvtypes "github.com/cosmos/interchain-security/v3/x/ccv/types"
+)
+
+// Migrator is a struct for handling in-place store migrations.
+type Migrator struct {
+	providerKeeper Keeper
+	paramSpace     paramtypes.Subspace
+}
+
+// NewMigrator returns a new Migrator.
+func NewMigrator(providerKeeper Keeper, paramSpace paramtypes.Subspace) Migrator {
+	return Migrator{providerKeeper: providerKeeper, paramSpace: paramSpace}
+}
+
+// Migrate2to3 migrates x/ccvprovider state from consensus version 2 to 3.
+func (m Migrator) Migrate2to3(ctx sdktypes.Context) error {
+	return m.providerKeeper.MigrateQueuedPackets(ctx)
+}
+
+func (k Keeper) MigrateQueuedPackets(ctx sdktypes.Context) error {
+	for _, consumer := range k.GetAllConsumerChains(ctx) {
+		slashData, vscmData := k.GetAllThrottledPacketData(ctx, consumer.ChainId)
+		if len(slashData) > 0 {
+			ctx.Logger().Error(fmt.Sprintf("slash data being dropped: %v", slashData))
+		}
+		for _, data := range vscmData {
+			k.HandleVSCMaturedPacket(ctx, consumer.ChainId, data)
+		}
+		k.DeleteThrottledPacketDataForConsumer(ctx, consumer.ChainId)
+	}
+	return nil
+
+}
+
+// Pending packet data type enum, used to encode the type of packet data stored at each entry in the mutual queue.
+// Note this type is copy/pasted from throttle v1 code.
+const (
+	slashPacketData byte = iota
+	vscMaturedPacketData
+)
+
+// GetAllThrottledPacketData returns all throttled packet data for a given consumer chain, only used for 2 -> 3 migration.
+// Note this method is adapted from throttle v1 code.
+func (k Keeper) GetAllThrottledPacketData(ctx sdktypes.Context, consumerChainID string) (
```

Comment:
Not optimal that we have to keep these methods around from throttling v1. But it's the only way to test a migration until #1296 is implemented